### ISSUE TITLE
Details about training setting

### ISSUE THREAD

**Issue by ymchen7** [Zheng-Chong/CatVTON#14]
Good work for the design of such simple vton pipeline. 

I have tried to train CatVTON on vitonhd dataset, but the result is a little blurry as shown below.   (38k iteration batchsize 8x32 512x384 resolution input, only attention parameters are trained)
<img width="768" alt="image" src="https://github.com/user-attachments/assets/6a39a5f8-e902-4d42-bd98-b3d0d54a57f3">

I'm wondering is there any specific setting or trick in the loss part, for example how to compute the loss? (i.e. compute loss of  latents of human images or the concat latents. )

I also noticed the training loss is relatively small at the beginning of the training, is this normal?

Epoch 0, step 0, step_loss: 0.06322, data_time: 2.104, time: 4.421
Epoch 0, step 1, step_loss: 0.04681, data_time: 0.058, time: 2.126
Epoch 0, step 2, step_loss: 0.06814, data_time: 0.058, time: 2.124
Epoch 0, step 3, step_loss: 0.03120, data_time: 0.064, time: 2.139
Epoch 0, step 4, step_loss: 0.02966, data_time: 0.059, time: 2.132
Epoch 0, step 5, step_loss: 0.03977, data_time: 0.059, time: 2.132
Epoch 0, step 6, step_loss: 0.05645, data_time: 0.059, time: 2.133

**Comment by Zheng-Chong** [comment]
It looks like there is nothing wrong with your visualization results and loss.
The loss is calculated on the concat latents. The training details should have been explained in the paper, and no other special tricks in loss part.

**Comment by BarryyyDrew** [comment]
Is it trainable？Could you please tell me how to train this model?

**Comment by franciszzj** [comment]
@ymchen7 @Zheng-Chong 
I trained on both only loss for person latents and loss for concat latents, and found that the results are similar.
Because the garment is provided as a condition, it is very easy for the model to learn from it. The part that really makes a difference is the loss of the person latents.

**Comment by Zheng-Chong** [comment]
> @ymchen7 @Zheng-Chong I trained on both only loss for person latents and loss for concat latents, and found that the results are similar. Because the garment is provided as a condition, it is very easy for the model to learn from it. The part that really makes a difference is the loss of the person latents.

That makes sense.

**Comment by Abhij-ma** [comment]
@ymchen7 Hello sir can you tell me how I can train this model on my machine

**Comment by HamnaAkram** [comment]
@ymchen7 how did you create the training pipeline? considering no details are given in the code.

**Comment by awais-nayyar** [comment]
@ymchen7 I also want to know, can you please guide us how can we train this model on our machine?

**Comment by catsled** [comment]
![image](https://github.com/user-attachments/assets/fc55079d-1390-40ee-b11d-bb1895acd909)

训练40ksteps, batch_size=8, 4张amd显卡，请问这个结果是否与训练步数过少有关

**Comment by catsled** [comment]
> Good work for the design of such simple vton pipeline.
> 
> I have tried to train CatVTON on vitonhd dataset, but the result is a little blurry as shown below. (38k iteration batchsize 8x32 512x384 resolution input, only attention parameters are trained) <img alt="image" width="768" src="https://private-user-images.githubusercontent.com/25624964/354459723-6a39a5f8-e902-4d42-bd98-b3d0d54a57f3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAzNTY4MzAsIm5iZiI6MTczMDM1NjUzMCwicGF0aCI6Ii8yNTYyNDk2NC8zNTQ0NTk3MjMtNmEzOWE1ZjgtZTkwMi00ZDQyLWJkOTgtYjNkMGQ1NGE1N2YzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDMxVDA2MzUzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk4ZWM0NjA4ZDljNjg2YWY0ODAwZDExY2YwY2EzNGJjYmY2ZjhmY2E2OWM0MWJjNjQxYmU2YTcyOTQyN2QwYzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.CoqlgOWQKxkIIG9U6dxrI4KQJkG_QA6JB_O_VfXwfqU">
> 
> I'm wondering is there any specific setting or trick in the loss part, for example how to compute the loss? (i.e. compute loss of latents of human images or the concat latents. )
> 
> I also noticed the training loss is relatively small at the beginning of the training, is this normal?
> 
> Epoch 0, step 0, step_loss: 0.06322, data_time: 2.104, time: 4.421 Epoch 0, step 1, step_loss: 0.04681, data_time: 0.058, time: 2.126 Epoch 0, step 2, step_loss: 0.06814, data_time: 0.058, time: 2.124 Epoch 0, step 3, step_loss: 0.03120, data_time: 0.064, time: 2.139 Epoch 0, step 4, step_loss: 0.02966, data_time: 0.059, time: 2.132 Epoch 0, step 5, step_loss: 0.03977, data_time: 0.059, time: 2.132 Epoch 0, step 6, step_loss: 0.05645, data_time: 0.059, time: 2.133

i have also the same issue, the color is not correct for many cloth, such the light red -> deep red. do you find some methods to solve it?

**Comment by Harsh-Kesharwani** [comment]
@ymchen7 can you please verify below computation is correct or not for mask based training.
```
    def compute_loss(self, batch: Dict[str, torch.Tensor]) -> torch.Tensor:
        """Compute MSE loss for denoising with DREAM strategy"""
        person_images = batch['person'].to(self.device)
        cloth_images = batch['cloth'].to(self.device)
        masks = batch['mask'].to(self.device)

        concat_dim=-2
        # Prepare inputs to Tensor
        image, condition_image, mask = check_inputs(person_images, cloth_images, masks, self.width, self.height)
        image = prepare_image(person_images).to(self.device, dtype=self.weight_dtype)
        condition_image = prepare_image(cloth_images).to(self.device, dtype=self.weight_dtype)
        mask = prepare_mask_image(masks).to(self.device, dtype=self.weight_dtype)
        # Mask image
        masked_image = image * (mask < 0.5)

        with torch.cuda.amp.autocast(enabled=self.use_mixed_precision):
        # VAE encoding
            person_latent=compute_vae_encodings(image, self.encoder, self.device)
            masked_latent = compute_vae_encodings(masked_image, self.encoder, self.device)
            condition_latent = compute_vae_encodings(condition_image, self.encoder, self.device)
            mask_latent = torch.nn.functional.interpolate(mask, size=masked_latent.shape[-2:], mode="nearest")
            del image, mask, condition_image

            # Apply CFG dropout to garment latent
            condition_latent = self._apply_cfg_dropout(condition_latent)
            
            # Concatenate latents
            masked_latent_concat = torch.cat([masked_latent, condition_latent], dim=concat_dim)
            mask_latent_concat = torch.cat([mask_latent, torch.zeros_like(mask_latent)], dim=concat_dim)
            target_latents=torch.cat([person_latent, condition_latent], dim=concat_dim)

            noise=randn_tensor(
                masked_latent_concat.shape,
                generator=self.generator,
                device=masked_latent_concat.device,
                dtype=self.weight_dtype,
            )

            timesteps = torch.randint(1, 1000, size=(1,))[0].long().item()
            timesteps_embedding=get_time_embedding(timesteps).to(self.device)

            # Add noise to latents
            noisy_latents = self.scheduler.add_noise(masked_latent_concat, timesteps, noise)

            inpainting_latent_model_input = torch.cat([
                noisy_latents, 
                mask_latent_concat, 
                masked_latent_concat
            ], dim=1).to(self.device, dtype=self.weight_dtype)

            # DREAM strategy implementation
            if self.dream_lambda > 0:
                
                # Get initial noise prediction
                with torch.no_grad():
                    epsilon_theta = self.diffusion(
                    inpainting_latent_model_input,
                    timesteps_embedding
                )

                # print(f"Predicted noise shape: {epsilon_theta.shape}")
                
                # Apply DREAM: zˆt = √αt*z0 + √(1-αt)*(ε + λ*εθ)
                alphas_cumprod = self.scheduler.alphas_cumprod.to(device=self.device, dtype=self.weight_dtype)
                sqrt_alpha_prod = alphas_cumprod[timesteps] ** 0.5
                sqrt_one_minus_alpha_prod = (1 - alphas_cumprod[timesteps]) ** 0.5
                
                # Reshape for broadcasting
                sqrt_alpha_prod = sqrt_alpha_prod.view(-1, 1, 1, 1)
                sqrt_one_minus_alpha_prod = sqrt_one_minus_alpha_prod.view(-1, 1, 1, 1)
                
                # DREAM noise combination
                dream_noise = noise + self.dream_lambda * epsilon_theta
                dream_noisy_latents = sqrt_alpha_prod * target_latents + sqrt_one_minus_alpha_prod * dream_noise
                dream_model_input = torch.cat([
                    dream_noisy_latents, 
                    mask_latent_concat, 
                    masked_latent_concat
                ], dim=1)

                predicted_noise= self.diffusion(
                    dream_model_input,
                    timesteps_embedding
                )
                # DREAM loss: |(ε + λεθ) - εθ(ẑt, t)|²
                loss = F.mse_loss(predicted_noise, dream_noise)
            else:
                # Standard training without DREAM
                predicted_noise = self.diffusion(
                    inpainting_latent_model_input,
                    timesteps_embedding,
                )

                # Standard MSE loss
                loss = F.mse_loss(predicted_noise, noise)
            
        return loss
```
